packages/react-form-renderer/src/files/ui-state-reducer.js

```diff
@@ -0,0 +1,92 @@
+const validAttributes = [
+  'visible',
+  'disabled',
+  'hidden',
+  'enabled',
+  'icon',
+  'help',
+  'colorBar',
+  'required',
+  'max',
+  'min',
+];
+
+//enabled and hidden are not used internally, but accepted as input from the conditions object
+const inversedTypes = {
+  enabled: 'disabled',
+  hidden: 'visible',
+};
+
+const uiStateReducer = (state, action) => {
+  switch (action.type) {
+    case 'addUIState': {
+      const {source, uiState} = action;
+
+      Object.entries(uiState).forEach(([key, item]) => {
+        //Create the item object for this item if it doesn't exist
+        if (!state.fields[key]) state.fields[key] = {};
+
+        validAttributes.forEach(type => {
+          //Don't add uiTypes if they don't exist in the dispatched message
+          if (item[type] === undefined) return;
+
+          //Handle inversed types (disabled/enabled, visible/hidden)
+          const inversedType = inversedTypes[type];
+          const value = inversedType ? !item[type] : item[type];
+          type = inversedType || type;
+
+          if (!state.fields[key][type]) {
+            //If this type doesn't exists for this item, we create a new array with only this source. No need to search fot the source
+            state.fields[key][type] = [{source, value}];
+          } else {
+            const index = state.fields[key][type].findIndex(item => item.source === source);
+            if (index !== -1) {
+              //If this type for this item from this source existed, update the state (value could change if condition went from "then" to "else")
+              state.fields[key][type][index].value = value;
+            } else {
+              //Otherwise, add the state from this source at the begining of the array (i.e. this will supress result from other sources)
+              state.fields[key][type].unshift({source, value});
```

Comment:
the `{...foo}` is not a deep clone. If we want to change nested non-primitive types we have to clone them as well. Using `push`, `shift` etc. will not create a new instance and that might not trigger React render function. For example:

```jsx
x = [1,2,3]
a = {b: {c: x}}
b = {...a}
a.b.c === b.b.c // results in true and that will not trigger new render
```

Reply:
Fixed in 12ce210